### PR TITLE
Convert middleware to TypeScript, plus some other things.

### PR DESCRIPTION
**Pull request by aomarks**
- Convert middleware to TypeScript.
- Remove unnecessary files from both NPM packages.
- Upgrade to TypeScript 3, and tweak some configs.
- Replace eslint with tslint and fix warnings.
- Add CHANGELOG to middleware.

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -0,0 +1,17 @@
+src/
+middleware/
+build/test/
+test-resources/
+thirdfonts/
```
Comment (aomarks) [inline .npmignore]: @samuelli Is this `thirdfonts/` directory needed at all?

File: middleware/src/test/middleware-test.ts [hunk 2]
```diff
@@ -63,12 +58,12 @@ const human = 'Chrome';
...
 /**
  * GET a URL with the given user agent.
- * @param {string} userAgent The user agent string.
- * @param {string} host The host part of the URL.
- * @param {string} path The path part of the URL.
- * @return {Promise<!Object>} Promise of the GET response.
+ * @param userAgent The user agent string.
+ * @param host The host part of the URL.
+ * @param path The path part of the URL.
+ * @return Promise of the GET response.
```
Comment (samuelli) [inline middleware/src/test/middleware-test.ts]: jsdoc can be removed

Comment (aomarks) [reply middleware/src/test/middleware-test.ts]: > jsdoc can be removed

Done.

**Comment by aomarks** [comment]
> Not any more, it can be removed entirely.

Done.

**Pull request merged**